flink-runtime/src/main/java/org/apache/flink/runtime/failurerate/TimestampBasedFailureRater.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.failurerate;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.runtime.execution.ThresholdExceedException;
+
+import java.util.ArrayDeque;
+import java.util.Queue;
+import java.util.function.Supplier;
+
+/**
+ * A timestamp queue based failure rater implementation.
+ */
+public class TimestampBasedFailureRater implements ThresholdMeter {
```

Comment:
There actually a three-layer concept here. ThresholdMeter is a Meter that has a checkAgainstThreshold interface. The TimestampBasedThresholdMeter is a concrete implementation. The failure rate is actually a customized TimestampBasedThresholdMeter, as we put the fromConfiguration and exception message in the class. I am good with split into three layers. But as it is the only use case, probably it will be better to keep it as it is?

Reply:
I believe the three layers you mean are the following.
1. Meter (interface)
2. ThresholdMeter (interface)
3. TimestampBasedThresholdMeter (implementation)

My questions is, why do we need 2) and 3) separately? What if we consider `ThresholdMeter` an implementation of `Meter` and have the following two layers?
1. Meter (interface)
2. ThresholdMeter (implementation)

I'm proposing this because：
* There's no other implementation of `ThresholdMeter` except for `TimestampBasedThresholdMeter`.
* Looking into the implementation of `TimestampBasedFailureRater`, there's nothing specifically related to *Failure*, and shouldn't the concept *Meter* already *TimestampBased*?

Reply:
Yes, agree we don't need to have three layers. The latest PR contains only two layers, Meter and TimestampBasedFailureRater. Why I don't want to use ThresholdMeter as the implementation class name? This two method checkThreshold and fromConfiguration has container failure related error message and configuration key. It is not generic enough to put into ThresholdMeter.

Reply:
I believe a clear abstraction of concepts should be prior to the implementation details (what the error messages say, how the factories should look like, etc). Compared with `Meter`, conceptually the only additional function this implementation class provides is `checkAgainstThreshold`.

Concerning the implementation issues you mentioned, they should be adjusted with respect to the underlying concepts.
* `checkAgainstThreshold` can only say that current rate `x` has reached the threshold `y`, and we can mention that the rate is for starting container failure in `ActiveResourceManager` for logging.
* We can change `fromConfiguration(Configuration config)` to `withMinuteThreshold(double threshold)` and parse the configuration option in `ActiveResourceManagerFactory`.